File: hola-mundo-backend-2022-grupo-7/src/main/kotlin/trabajoPracticoAlgoritmos2/Criterio.kt
package trabajoPracticoAlgoritmos2

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo

@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type")
@JsonSubTypes(
    JsonSubTypes.Type(value = Relajado::class, name = "Relajado"),
    JsonSubTypes.Type(value = Precavido::class, name = "Precavido"),
    JsonSubTypes.Type(value = Localista::class, name = "Localista"),
    JsonSubTypes.Type(value = Soniador::class, name = "Soniador"),
    JsonSubTypes.Type(value = Activo::class, name = "Activo"),
    JsonSubTypes.Type(value = Exigente::class, name = "Exigente")
)
open class Criterio() {

    open lateinit var tipo:String

    open fun aceptaItinerario(usuario: Usuario, itinerario: Itinerario): Boolean {
        return false
    }
}

class Relajado(
    override var tipo: String = "Relajado"
) : Criterio() {



    override fun aceptaItinerario(usuario: Usuario, itinerario: Itinerario): Boolean {
        return true
    }

}

class Precavido(
    override var tipo: String = "Precavido"
) : Criterio() {
    override fun aceptaItinerario(usuario: Usuario, itinerario: Itinerario): Boolean {
        return (usuario.amigos.flatMap { it.destinosVisitados }).contains(itinerario.destino) or (usuario.conoceDestino(
            itinerario.destino
        ))
    }
}

class Localista(
    override var tipo: String = "Localista"
) : Criterio() {
    override fun aceptaItinerario(usuario: Usuario, itinerario: Itinerario): Boolean {
        return itinerario.destino.esLocal()
    }
}

class Soniador(
    override var tipo: String = "Soniador"
) : Criterio() {
    override fun aceptaItinerario(usuario: Usuario, itinerario: Itinerario): Boolean {
        val aux = usuario.destinosDeseados.maxOf { it.calcularCostoBase() }
        return (aux < itinerario.destino.calcularCostoBase()) or (usuario.destinosDeseados.contains(itinerario.destino))
    }
}

class Activo(
    override var tipo: String = "Activo"
) : Criterio() {
    override fun aceptaItinerario(usuario: Usuario, itinerario: Itinerario): Boolean {
        return itinerario.diasDeActividad.all { !it.estaVacio() }

    }
}

class Exigente(
    var dificultad: Dificultad,
    var porcentaje: Double,
    override var tipo: String = "Exigente"
    ) : Criterio() {

    override fun aceptaItinerario(usuario: Usuario, itinerario: Itinerario): Boolean {
        return itinerario.porcentajeDificultadDeActividad(dificultad) >= porcentaje
    }
}

File: Angular/src/Dominio/Criterio/Criterio.ts
import { Dificultad } from 'src/Dominio/Actividades/Actividades';
import { Itinerario } from "../Itinerarios/Itinerarios";
import { Usuario } from "../Usuario/Usuario";

export type CriterioJson = {

    type: string,
    dificultad:Dificultad,
    porcentaje:number,

}

export class Criterio{

    type:string = ""
    aceptaItinerario(usuario: Usuario, itinerario: Itinerario): boolean{
        return false
    }

    static fromJson(criterioJson:CriterioJson):Criterio{

        switch(criterioJson.type){

            case "Relajado" : {
                return new Relajado()
            }
            case "Precavido" : {
                return new Precavido()
            }
            case "Localista" : {
                return new Localista()
            }
            case "Soniador" : {
                return new Soniador()
            }
            case "Activo" : {
                return new Activo()
            }
            case "Exigente" : {
                return new Exigente(criterioJson.dificultad,criterioJson.porcentaje)
            }
        }
        throw new Error("Tipo Invalido")
    }
/*
    toJSON(criterio:Criterio):CriterioJson{
        return {
            type: this.type,
            dificultad: this.dificultad,
            porcentaje: this.porcentaje,
        }
    }
    */
}

export class Relajado extends Criterio{
    override type = "Relajado"

    override aceptaItinerario(usuario: Usuario, itinerario: Itinerario): boolean {
        return true
    }

}

export class Precavido extends Criterio{
    override type = "Precavido"

    override aceptaItinerario(usuario: Usuario, itinerario: Itinerario): boolean {
        return usuario.amigos.flatMap((amigo)=>{return amigo.destinosVisitados}).includes(itinerario.destino)  || usuario.conoceDestino(itinerario.destino)
    }

}

export class Localista extends Criterio{
    override type = "Localista"

    override aceptaItinerario(usuario: Usuario, itinerario: Itinerario): boolean {
        return itinerario.destino.esLocal()
    }

}

export class Soniador extends Criterio {
    override type = "Soniador"

    override aceptaItinerario(usuario: Usuario, itinerario: Itinerario): boolean {

        const costosDeLosDestinosDelUsuario = usuario.destinosDeseados.map((destino)=>{return destino.calcularCostoBase()})
        const destinoDeseadoMasCaro = Math.max(...costosDeLosDestinosDelUsuario)

        return destinoDeseadoMasCaro < itinerario.destino.calcularCostoBase() || usuario.destinosDeseados.includes(itinerario.destino)

    }

}

export class Activo extends Criterio {
    override type = "Activo"

    override aceptaItinerario(usuario: Usuario, itinerario: Itinerario): boolean {
        return itinerario.diasDeActividad.every((dia)=>{return !dia.estaVacio()})
    }

}

export class Exigente extends Criterio {
    override type = "Exigente"

    dificultad:Dificultad
    porcentaje:number

    constructor(_dificultad:Dificultad, _porcentaje:number){
        super()
        this.dificultad = _dificultad
        this.porcentaje = _porcentaje
    }

    override aceptaItinerario(usuario: Usuario, itinerario: Itinerario): boolean {
        return itinerario.porcentajeDificultadDeActividad(this.dificultad) >= this.porcentaje
    }

}